src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (finalDelimiterNeedsSemicolon)
+            {
+                if (!StatementClosingDelimiterExists(currentNode, ref lastDelimiterSpan))
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
+                    nextCommandHandler();
+                    return;
+                }
+            }
+
+            // if you haven't found any enclosures, put semicolon at end of statement
+            if (lastDelimiterSpan == default)
+            {
+                lastDelimiterSpan = currentNode.Span;
+            }
+
+            // Move to space after the last delimiter
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, lastDelimiterSpan.End, currentNode.Kind())));
+            nextCommandHandler();
+        }
+
+        private static SyntaxNode GetEnclosingArgumentList(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts)
+        {
+            while (currentNode.Kind() != SyntaxKind.ArgumentList
+                && currentNode.Kind() != SyntaxKind.ArrayRankSpecifier)
+            {
+                if (currentNode.Kind() == SyntaxKind.InterpolatedStringExpression || currentNode.Kind() == SyntaxKind.StringLiteralExpression)
+                {
+                    return null;
+                }
+
+                if (currentNode == null
+                    || syntaxFacts.IsStatement(currentNode)
+                    || currentNode.Kind() == SyntaxKind.VariableDeclaration)
+                {
+                    return null;
+                }
+
+                currentNode = currentNode.Parent;
+                if (currentNode == null)
+                {
+                    return null;
+                }
+            }
+
+            // now we're in an argument list, so return the enclosing statement
+            while (!syntaxFacts.IsStatement(currentNode) && !(currentNode.Kind() == SyntaxKind.VariableDeclaration))
+            {
+                currentNode = currentNode.Parent;
+                if (currentNode == null)
+                {
+                    return null;
+                }
+            }
+
+            return currentNode;
+        }
+
+        private static bool IsCaretAtEndOfLine(SnapshotPoint caret)
+        {
+            return caret.Position == caret.GetContainingLine().End;
+        }
+
+        private static SyntaxToken GetToken(SyntaxNode root, int caretPosition, SnapshotPoint caret)
+        {
+            if (IsCaretAtEndOfLine(caret) && caretPosition > 0)
+            {
+                return root.FindToken(caretPosition - 1);
+            }
+            else
+            {
+                return root.FindToken(caretPosition);
+            }
+        }
+
+        private static bool SemiColonIsMissing(SyntaxNode currentNode)
+        {
+            switch (currentNode.Kind())
```

Comment:
it feels like this can be written as:

```c#
var children = currentNode.ChildNodesAndTokens();
if (children.Count == 0) return false;
var last = children.Last();
return last.Kind() == SemicolonToken && last.IsMissing;
```

Reply:
➡️ It might be possible, but I don't mind leaving this expanded for now. It helps picture all the cases that might be of interest. I'll file a follow-up issue to simplify this once we address additional statement completion cases.

Reply:
I'm adding an assertion to the `default` case that the list is complete.

Reply:
tiny nit: consider making Colon lowercase

Reply:
I'm ok with a followup.  But this really does feel like it would be simpler in another form.  Effectively we're trying to say "if this is a construct whose last token is a non-optional semicolon, and that semicolon is missing, then blah blah".  So rather than having to enumerate, we can actually just enforce this form unilaterally.  But no need to do that in this PR :)

Reply:
Unresolved to make sure the follow-up issue gets filed.

Reply:
Here is the follow-up issue: https://github.com/dotnet/roslyn/issues/32163